src/ts/services/keyring-vault.ts

```diff
@@ -0,0 +1,171 @@
+import { KeyringInstance, KeyringPair, KeyringPair$Json } from '@polkadot/keyring/types'
+import Keyring from '@polkadot/keyring'
+import { Prefix } from '@polkadot/keyring/address/types'
+import { LocalStore } from './local-store'
+import { mnemonicGenerate, cryptoWaitReady, mnemonicValidate } from '@polkadot/util-crypto'
+import t from './i18n'
+
+class KeyringVault {
+
+  private _vaultKey: string = 'speckle-vault'
+  private _keyring?: KeyringInstance
+  private _password?: string
+  private _mnemonic?: string
+
+  private get keyring (): KeyringInstance {
+    if (this._keyring) {
+      return this._keyring
+    }
+    throw new Error(`Keyring is not initialised yet`)
+  }
+
+  unlock (password: string, addressPrefix?: Prefix): Promise<Array<KeyringPair$Json>> {
+    if (this.isUnlocked()) throw new Error(t('error.wallet.unlocked'))
+    if (!password.length) throw new Error(t('error.password'))
+    // this will be redundant if we have polkadot js api initialisation
+    return cryptoWaitReady().then(async () => {
+      this._keyring = new Keyring({ addressPrefix, type: 'sr25519' })
+      let vault = await LocalStore.get(this._vaultKey)
+      if (vault) {
+        let accounts = Object.values(vault[this._vaultKey])
+        try {
+          accounts.forEach((account) => {
+            let pair = this.keyring.addFromJson(account as KeyringPair$Json)
+            pair.decodePkcs8(password)
+            this.keyring.addPair(pair)
+          })
+          this._password = password
+          return accounts as Array<KeyringPair$Json>
+        } catch (e) {
+          this.keyring.getPairs().forEach((pair) => {
+            this.keyring.removePair(pair.address())
+          })
+          throw new Error(t('error.password'))
+        }
+      }
+      this._password = password
+      return []
+    })
+  }
+
+  getAccounts (): Array<KeyringPair$Json> {
+    if (this.isLocked()) throw new Error(t('error.wallet.locked'))
+    let accounts = new Array<KeyringPair$Json>()
+    this.keyring.getPairs().forEach((pair) => {
+      accounts.push(pair.toJson(this._password))
+    })
+    return accounts
+  }
+
+  lock () {
+    this._password = undefined
+    this._mnemonic = undefined
+    this._keyring = undefined
+  }
+
+  generateMnemonic (): string {
+    if (this.isLocked()) {
+      throw new Error(t('error.wallet.locked'))
+    }
+    this._mnemonic = mnemonicGenerate()
+    return this._mnemonic
+  }
+
+  createAccount (mnemonic: string, accountName: string): Promise<KeyringPair$Json> {
+    if (this.isLocked()) throw new Error(t('error.wallet.locked'))
+    if (this._mnemonic !== mnemonic) throw new Error(t('error.mnemonic.unmatched'))
+    let pair = this.keyring.addFromMnemonic(mnemonic, { name: accountName })
+    this._mnemonic = undefined
+    return this.saveAccount(pair).then((keyringPair$Json: KeyringPair$Json) => {
```

Comment:
i can see we have used `
return this.saveAccount(pair).then((keyringPair$Json: KeyringPair$Json) => {
      return keyringPair$Json
    })
` multiply times. We probably can put it to a seperate func

Reply:
Just realised that saveAccount itself already returns a Promise that resolves to the KeyringPair$Json, the then parts can be omitted